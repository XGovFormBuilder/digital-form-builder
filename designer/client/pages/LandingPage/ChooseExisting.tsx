import React, { Component } from "react";
import * as formConfigurationApi from "../../load-form-configurations";
import { i18n } from "../../i18n";
import { withRouter } from "react-router-dom";
import { BackLink } from "../../components/BackLink";
import "./LandingPage.scss";

type Props = {
  history: any;
};

type State = {
  configs: { Key: string; DisplayName: string }[];
  loading?: boolean;
};

export class ChooseExisting extends Component<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      configs: [],
      loading: true,
    };
  }

  componentDidMount() {
    formConfigurationApi.loadConfigurations().then((configs) => {
      this.setState({
        loading: false,
        configs,
      });
    });
  }

  selectForm = async (form) => {
    try {
      const response = await window.fetch("/api/new", {
        method: "POST",
        body: JSON.stringify({
          selected: { Key: form },
          name: "",
        }),
        headers: {
          Accept: "application/json",
          "Content-Type": "application/json",
        },
      });
      const responseJson = await response.json();
      this.props.history.push(`/designer/${responseJson.id}`);
    } catch (e) {
      console.error(e);
    }
  };

  goBack = (event) => {
    event.preventDefault();
    this.props.history.goBack();
  };

  render() {
    const configs = this.state.configs || [];
    const hasEditableForms = configs.length > 0;
    if (this.state.loading) {
      return <p>Loading ...</p>;
    }

    const formTable = configs.map((form) => (
      <tr className="govuk-table__row" key={form.Key}>
        <td className="govuk-table__cell">
          <a
            className="govuk-link"
            href="#"
            onClick={(e) => {
              e.preventDefault();
              this.selectForm(form.Key);
            }}
          >
            {form.DisplayName}
          </a>
        </td>
      </tr>
    ));

    return (
      <div className="new-config">
        <div>
          <BackLink onClick={this.goBack}>
            {i18n("Back to previous page")}
          </BackLink>

          <h1 className="govuk-heading-l">
            {i18n("landingPage.existing.select")}
          </h1>

          <div className="govuk-grid-row form-grid">
            <div className="govuk-grid-column-two-thirds">
              <table className="govuk-table">
                <thead className="govuk-table__head">
                  <tr className="govuk-table__row">
                    <th scope="col" className="govuk-table__header">
                      Form name
                    </th>
                  </tr>
                </thead>
                <tbody className="govuk-table__body">
                  {hasEditableForms ? (
                    <>{formTable}</>
                  ) : (
                    <tr className="govuk-table__row">
                      <td className="govuk-table__cell table__cell__noborder">
                        {i18n("landingPage.existing.noforms")}
                      </td>
                    </tr>
                  )}
                </tbody>
              </table>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

export default withRouter(ChooseExisting);
