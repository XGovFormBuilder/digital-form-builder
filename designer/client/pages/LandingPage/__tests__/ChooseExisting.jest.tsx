import React from "react";
import { ChooseExisting } from "../ChooseExisting";
import {
  render,
  cleanup,
  fireEvent,
  screen,
  waitFor,
} from "@testing-library/react";
import {
  server,
  rest,
  mockedFormConfigurations,
} from "../../../../test/testServer";

describe("ChooseExisting", () => {
  afterEach(() => jest.resetAllMocks());
  afterEach(cleanup);

  test("no existing configurations", async () => {
    server.resetHandlers(
      rest.get("/api/configurations", (_req, res, ctx) => {
        return res(ctx.json([]));
      })
    );

    const push = jest.fn();
    const history = { push: push };
    const { asFragment } = render(<ChooseExisting history={history} />);
    expect(await screen.findByText(/Form name/i)).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  test("with existing configurations", async () => {
    const push = jest.fn();
    const history = { push: push };
    const { asFragment } = render(<ChooseExisting history={history} />);
    expect(await screen.findByText(/Form name/i)).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  test("should post to server and redirect to a new page on choosing a form", async () => {
    server.use(
      rest.post("/api/new", (req, res, ctx) => {
        return res(ctx.json({ id: "somekey", previewUrl: "" }));
      })
    );
    const push = jest.fn();
    const history = { push: push };

    render(<ChooseExisting history={history} />);
    expect(await screen.findByText(/Form name/i)).toBeInTheDocument();

    fireEvent.click(screen.getByText(mockedFormConfigurations[0].DisplayName));
    await waitFor(() => expect(push).toHaveBeenCalledTimes(1));
    expect(push).toBeCalledWith("/designer/somekey");
  });
});
