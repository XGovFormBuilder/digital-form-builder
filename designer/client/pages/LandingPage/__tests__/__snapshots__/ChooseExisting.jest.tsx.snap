// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChooseExisting no existing configurations 1`] = `
<DocumentFragment>
  <div
    class="new-config"
  >
    <div>
      <a
        class="back-link govuk-back-link"
        href="#0"
      >
        Back to previous page
      </a>
      <h1
        class="govuk-heading-l"
      >
        Select an existing form
      </h1>
      <div
        class="govuk-grid-row form-grid"
      >
        <div
          class="govuk-grid-column-two-thirds"
        >
          <table
            class="govuk-table"
          >
            <thead
              class="govuk-table__head"
            >
              <tr
                class="govuk-table__row"
              >
                <th
                  class="govuk-table__header"
                  scope="col"
                >
                  Form name
                </th>
              </tr>
            </thead>
            <tbody
              class="govuk-table__body"
            >
              <tr
                class="govuk-table__row"
              >
                <td
                  class="govuk-table__cell table__cell__noborder"
                >
                  You do not have any existing forms
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ChooseExisting with existing configurations 1`] = `
<DocumentFragment>
  <div
    class="new-config"
  >
    <div>
      <a
        class="back-link govuk-back-link"
        href="#0"
      >
        Back to previous page
      </a>
      <h1
        class="govuk-heading-l"
      >
        Select an existing form
      </h1>
      <div
        class="govuk-grid-row form-grid"
      >
        <div
          class="govuk-grid-column-two-thirds"
        >
          <table
            class="govuk-table"
          >
            <thead
              class="govuk-table__head"
            >
              <tr
                class="govuk-table__row"
              >
                <th
                  class="govuk-table__header"
                  scope="col"
                >
                  Form name
                </th>
              </tr>
            </thead>
            <tbody
              class="govuk-table__body"
            >
              <tr
                class="govuk-table__row"
              >
                <td
                  class="govuk-table__cell"
                >
                  <a
                    class="govuk-link"
                    href="#"
                  >
                    Not a feedback form
                  </a>
                </td>
              </tr>
              <tr
                class="govuk-table__row"
              >
                <td
                  class="govuk-table__cell"
                >
                  <a
                    class="govuk-link"
                    href="#"
                  >
                    My feedback form
                  </a>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
